File: 백준/Silver/1996. 지뢰 찾기/지뢰 찾기.py
import sys
data_temp = sys.stdin if sys.platform == 'linux' else open('입력.txt', 'r')
input_data = data_temp.read().splitlines()

def solution (data) :
  n = int(data.pop(0))
  matrix = list(map(lambda x : list(x), data))
  locationOfMines = {}
  
  for i in range(n) :
    for j in range(n) :
      if matrix[i][j].isdigit() : 
        locationOfMines['{}:{}'.format(i,j)] = '*'
        directions = [[-1,-1],[0,-1],[1,-1],[1,0],[1,1],[0,1],[-1,1],[-1,0]]
        for [dx,dy] in directions :
          nx, ny = dx + j, dy + i
          if verify(n, nx, ny) :
            temp = '{}:{}'.format(ny,nx)
            tempVerify = locationOfMines.get(temp, None)
            if tempVerify != '*' and tempVerify != 'M' : 
              locationOfMines[temp] = locationOfMines.get(temp, 0) + int(matrix[i][j])
              if locationOfMines[temp] > 9 : locationOfMines[temp] = 'M'
  
  result = [['0' for _ in range(n)] for _ in range(n)]
  
  for key, value in locationOfMines.items() :
    y,x = map(int, key.split(':'))
    result[y][x] = str(value)
    
  print('\n'.join(map(lambda x : ''.join(x),result)))
  
def verify (n,y,x) : 
  return 0 <= x < n and 0 <= y < n
  
solution(input_data)

File: 백준/Silver/4396. 지뢰 찾기/지뢰 찾기.js
const fs = require("fs");
const input = fs
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "입력.txt")
  .toString()
  .trim()
  .split("\n");

function solution(data) {
  const N = +data.shift();

  const mines = data.slice(0, N),
    dig = data.slice(N);

  const result = Array.from({ length: N }, () => {
    return Array.from({ length: N }, () => ".");
  });

  const verify = (x, y) => 0 <= x && x < N && 0 <= y && y < N;

  function findMines(x, y) {
    if (mines[y][x] === "*") return -1;

    const directions = [
      [-1, -1],
      [-1, 0],
      [-1, 1],
      [0, -1],
      [0, 1],
      [1, -1],
      [1, 1],
      [1, 0],
    ];

    let cnt = 0;
    directions.forEach(([dx, dy]) => {
      const [nx, ny] = [dx + x, dy + y];
      if (verify(nx, ny) && mines[ny][nx] === "*") cnt++;
    });

    return cnt;
  }

  let gameOverToggle = false;
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      if (dig[i][j] === "x") {
        const k = findMines(j, i);
        result[i][j] = k;
        if (k === -1) gameOverToggle = true;
      }
    }
  }

  function gameOver() {
    for (let i = 0; i < N; i++) {
      for (let j = 0; j < N; j++) {
        if (mines[i][j] === "*") result[i][j] = "*";
      }
    }
  }

  if (gameOverToggle) gameOver();
  console.log(result.map((el) => el.join("")).join("\n"));
}

solution(input);
